### PR TITLE
ARROW-11141: [Rust] Add basic Miri checks to CI pipeline

### PR DESCRIPTION
**Pull request by vertexclique**

### CODE REVIEW

**Comment by vertexclique** [comment]
Checks are not running in this PR.

**Comment by alamb** [comment]
> Checks are not running in this PR.


@vertexclique  I wonder if you could trigger the checks by making some minor change to the rust source files (e.g a comment update or something)

**Comment by vertexclique** [comment]
If that triggers that would be nice. Can you commit to this branch?

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -226,6 +226,37 @@ jobs:
...
+      - uses: actions/checkout@v2
+      - uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cargo/registry
+            ~/.cargo/git
+            target
+          key: ${{ runner.os }}-cargo-miri-${{ hashFiles('**/Cargo.lock') }}
+      - name: Install
+        uses: actions-rs/toolchain@v1
```
Comment (jorgecarleitao) [inline .github/workflows/rust.yml:246]: This action is blocked by infra. Could you try using one of the commands on the other steps?

Comment (vertexclique) [reply .github/workflows/rust.yml]: Cool, I have changed it. Now it has started.

**Comment by vertexclique** [comment]
@alamb @jorgecarleitao checks are working right now.

**Comment by alamb** [comment]
@vertexclique  -- this is really cool. Thank you

So now I notice that the MIRI check is flagging what is presumably an error. Is that something you are planning to fix? If not I might be able to find some time to do so

https://github.com/apache/arrow/pull/9261/checks?check_run_id=1729267740

```
test array::builder::tests::test_string_dictionary_builder_with_reserved_null_value ... ok
error: Undefined Behavior: trying to reborrow for SharedReadWrite at alloc8703298, but parent tag <23938158> does not have an appropriate item in the borrow stack
    --> arrow/src/array/builder.rs:1460:9
     |
1460 |         self.field_anys[i].downcast_mut::<T>()
     |         ^^^^^^^^^^^^^^^^^^ trying to reborrow for SharedReadWrite at alloc8703298, but parent tag <23938158> does not have an appropriate item in the borrow stack
     |
     = help: this indicates a potential bug in the program: it performed an invalid operation, but the rules it violated are still experimental
     = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
             
     = note: inside `array::builder::StructBuilder::field_builder::<array::builder::GenericStringBuilder<i32>>` at arrow/src/array/builder.rs:1460:9
note: inside `array::builder::tests::test_struct_array_builder` at arrow/src/array/builder.rs:2827:30
    --> arrow/src/array/builder.rs:2827:30
     |
2827 |           let string_builder = builder
     |  ______________________________^
2828 | |             .field_builder::<StringBuilder>(0)
     | |______________________________________________^
note: inside closure at arrow/src/array/builder.rs:2813:5
    --> arrow/src/array/builder.rs:2813:5
     |
2813 | /     fn test_struct_array_builder() {
2814 | |         let string_builder = StringBuilder::new(4);
2815 | |         let int_builder = Int32Builder::new(4);
2816 | |
...    |
2892 | |         }
2893 | |     }
     | |_____^
     = note: inside `<[closure@arrow/src/array/builder.rs:2813:5: 2893:6] as std::ops::FnOnce<()>>::call_once - shim` at /usr/share/rust/.rustup/toolchains/nightly-2021-01-19-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
```

**Comment by vertexclique** [comment]
I want to merge the checks first then carry on with fixes in different prs as a cooperative effort to not making this pr and ticket complicated.

**Comment by alamb** [comment]
@vertexclique  merging the checks makes sense -- however, i think we need to figure out how to avoid github signaling a failure (red x) so that we can get clean CI runs

Screen Shot 2021-01-21 at 7 40 31 AM

Is there some way to instruct Miri to ignore failures or a known failure list?

**Comment by vertexclique** [comment]
@alamb fixed

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -226,6 +226,40 @@ jobs:
...
+      - name: Run Miri Checks
+        env:
+          RUST_BACKTRACE: full
+          RUST_LOG: 'trace'
+        run: |
+          export MIRIFLAGS="-Zmiri-disable-isolation"
+          cd rust
+          cargo miri setup
+          cargo clean
+          cargo miri test || true
```
Comment (alamb) [inline .github/workflows/rust.yml:261]:
```suggestion
          # Ignore MIRI errors until we can get a clean run
          cargo miri test || true
```

**Review by alamb — APPROVED** [review]
Thanks @vertexclique !

LGTM -- what do you think @jorgecarleitao  -- shall we start running this as part of CI as well? 

I think it is a great step forward.

**Review by paddyhoran — APPROVED** [review]

**Review by jorgecarleitao — APPROVED** [review]
Thanks a lot!!!!

**Pull request closed without merging**